### ISSUE TITLE
Make 'main' read-only

### ISSUE THREAD

**Issue by Anthchirp** [dials/dials#1353]
We recently had a bit of a bad experience where code that didn't pass tests on a pull request was merged. We propose to enable a setting in Github that commits can only be merged if tests pass. This will have the side-effect that all commits have to go through pull requests before landing on `master`.

While this may make low-risk quick fixes more cumbersome it would work quite nicely hand in hand with our towncrier/release notes process (which will be discussed in a separate issue/PR). It may also encourage cleaner packaging of changes and code reviews, and aid in backporting change sets to releases.

We will trial this workflow on github.com/xia2/xia2 for a month before making any changes to this repository. If that trial is deemed successful we would enable this for DIALS for a month and then reevaluate.

**Comment by Anthchirp** [comment]
So far we did find out that my comment is slightly wrong: not everything has to go through a pull request. It is fundamentally also possible to commit to a branch based off current `master`, push that branch, wait for the tests to pass on the branch and then fast-forward `master` to that commit.
However, this will only work if nobody else commits anything to `master` in the meantime. If `master` moved then a pull request needs to be opened for the branch.

Another point to note is that these restrictions will also apply to our regular Jenkins cleanup commits, so Jenkins may have to open pull requests instead.

**Comment by graeme-winter** [comment]
An anecdotal comment for this - 

When we brought in code formatting standards I found it annoying, but later realised it was good that the code always looks one way rather than different depending on who was the author. 

Enforcing this with black in pre-commit hooks was very annoying, but means that the diffs are clean every time, and you don't have to think about how to format code. For what it's worth I've started manually running clang-format on C++ code too. 

Forcing you to remove / fixed unused imports, unused variables etc. with flake8 in pre-commit hooks was very annoying, but is an excellent way to see that you've added a bug before a commit. 

Having recently merged a PR with failing tests which worked for me locally, I can really see this helping as it would catch things like this before merged and avoid "fixup" commits after merge, which makes picking across to release branches much easier. 

TL;DR - every time something like this has come along I have found it annoying and bossy and eventually a really good idea, which makes life slightly more annoying for me but makes working in the code base, reviewing pull requests and keeping track of what is going on far better for everyone involved.

**Comment by graeme-winter** [comment]
In discussion with @phyy-nx and others: this is not requiring that the pull request is reviewed - it is requiring that the tests have passed on the branch before merging (which could be via PR merged very swiftly). 

We can add a "merge-once-tests-pass" label to indicate that this is a change set which once would have been pushed direct to master and is not seeking review. 

We should add this to contributors documentation.

**Comment by Anthchirp** [comment]
[Outcome from today's core meeting](https://dials.github.io/kb/core/20200903): We will add this into the contributors documentation as a soft imposition (ie. "This is how we expect to work") without enforcing it at first, and then review after eg. a month.

**Comment by Anthchirp** [comment]
This has been introduced in https://dials.github.io/kb/core/20210617#status-checks-before-merging
and was reviewed after a trial period in https://dials.github.io/kb/core/20210715#status-checks-before-merging-on-dialsdials
and we decided to keep it.
